[KEYCLOAK-6116] - Get email attribute from 'subject alternative name' using X509 certificate

pedroigor opened the pull request:
Still need to work on tests. Need to figure out someway to include this scenario in X509 testsuite. It seems tests are really dependent on a single client certificate with no option to support different certificates.

pedroigor committed:
[KEYCLOAK-6116] - Get email attribute from 'subject alternative name' using X509 certificate

stianst reviewed (dismissed):
Needs tests to be added before merging

pedroigor committed:
[KEYCLOAK-6116] - Adding a OpenSSL CA Authority for testing

pedroigor committed:
[KEYCLOAK-6116] - Tests for X509 Subject Alternative Name Extension

pedroigor commented:
@stianst , I've added tests.

In order to be able to test this scenario and provide a better configuration for X509 related tests, I've also added a OpenSSL Certificate Authority. We should be able now to create certs more easily, using any extension we may need, check for certificate chains more easily, etc.

mposolda commented:
@pedroigor I am adding approve. Sorry it took so long. Will try to be better next time :)

I am adding @vmuzikar as another reviewer. AFAIK he is responsible for X509 testing on QE side. Even if he doesn't have concerns, it may be good that at least he is aware of those changes :) Especially PhantomJS setup changes and maybe other things too.

vmuzikar commented:
@pedroigor Will look into it first thing tomorrow. Would that be ok?

vmuzikar commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/AbstractX509AuthenticationTest.java:
```diff
@@ -100,6 +102,27 @@ protected boolean isImportAfterEachMethod() {
 
+    @BeforeClass
+    public static void onBeforeTestClass() {
+        if (System.getProperty("auth.server.container", "undefined").endsWith("wildfly")) {
```
This won't work wit auth-server-eap.

pedroigor replied:
True, forgot about eap ... I guess it should be enough to include a check of "eap" there ?

vmuzikar replied:
I believe something like that would work too:
```java
Boolean.parseBoolean(System.getProperty("auth.server.jboss"));
```

vmuzikar commented on testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/x509/X509BrowserLoginSubjectAltNameEmailTest.java:
```diff
@@ -0,0 +1,97 @@
+
+    @BeforeClass
+    public static void onBeforeTestClass() {
+        if (System.getProperty("auth.server.container", "undefined").endsWith("wildfly")) {
```
This won't work wit auth-server-eap.

vmuzikar reviewed (changes requested):
The X.509 tests won't work with EAP, i.e. RH-SSO. Could you please fix that?

Just my opinion: The new phantom.js config looks to me a bit unnecessary complicated - lots of copy-paste and commands just to change cert, key and passphrase. The main problem I see here is that we need to rewrite all the phantom.js configs just to use a client cert. Imagine we would need to change some global phantom.js config - then we'd need to think of all the tests which are using certs, and make the changes there as well. Couldn't we make some dedicated system props just for the cert, key and pass?
What do you think?

vmuzikar commented:
Also, should we update docs with this as well?

pedroigor commented:
@vmuzikar, here it is https://github.com/keycloak/keycloak-documentation/pull/334.

pedroigor committed:
[KEYCLOAK-6116] - Enabling tests for both jboss servers

pedroigor commented:
@vmuzikar, X509 tests are extending from ```AbstractX509AuthenticationTest```. 

The changes related with phantomjs is a workaround to overcome phantomjs config and execution in testsuite.

I would say that you could merge this PR and have a separated issue to think about how to properly handle these issues.

vmuzikar reviewed (approved):
@pedroigor Ok, LGTM then. Created https://issues.jboss.org/browse/KEYCLOAK-6792

pedroigor committed:
[KEYCLOAK-6116] - Removing references to phantomjs.cli.args

pedroigor commented:
@stianst, can we have your bless on this one ?
